sklearn/feature_selection/_sequential.py

```diff
@@ -0,0 +1,218 @@
+
+    See also
+    --------
+    RFE : Recursive feature elimination selection of the best number of
+          features
+    SelectFromModel : Feature selection based on importance weights
+    """
+    def __init__(self, estimator, *, n_features_to_select=None, forward=True,
+                 scoring=None, cv=5, n_jobs=None, pre_dispatch='2*n_jobs'):
+
+        self.estimator = estimator
+        self.n_features_to_select = n_features_to_select
+        self.forward = forward
+        self.pre_dispatch = pre_dispatch
+        self.scoring = scoring
+        self.cv = cv
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Learn the features to select.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training vectors.
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+
+        tags = self._get_tags()
+        X, y = self._validate_data(
+            X, y, accept_sparse="csc",
+            ensure_min_features=2,
+            force_all_finite=not tags.get('allow_nan', True),
+            multi_output=True
+        )
+
+        if self.n_features_to_select is None:
+            self.n_features_to_select_ = X.shape[1] // 2
+        else:
+            self.n_features_to_select_ = self.n_features_to_select
+
+        if not 1 <= self.n_features_to_select_ < X.shape[1]:
+            raise ValueError(
+                "n_features_to_select must be in [1, n_features - 1] = "
+                f"[1, {X.shape[1] - 1}]. Got {self.n_features_to_select_}."
+            )
+
+        cloned_estimator = clone(self.estimator)
+
+        # the current mask corresponds to the set of features:
+        # - that we have already *selected* if we do forward selection
+        # - that we have already *excluded* if we do backward selection
+        current_mask = set()
+        n_iterations = (self.n_features_to_select_ if self.forward
+                        else X.shape[1] - self.n_features_to_select_)
+        for _ in range(n_iterations):
+            new_feature_idx = self._get_best_new_feature(cloned_estimator, X,
+                                                         y, current_mask)
+            current_mask.add(new_feature_idx)
+
+        # transform the mask into a proper boolean mask of selected features
+        selected_features = (current_mask if self.forward
+                             else set(range(X.shape[1])) - current_mask)
+        self.support_ = np.zeros(X.shape[1], dtype=bool)
+        self.support_[list(selected_features)] = True
+
+        return self
+
+    def _get_best_new_feature(self, estimator, X, y, current_mask):
+        # Return the best new feature to add to the current_mask, i.e. return
+        # the best new feature to add (resp. remove) when doing forward
+        # selection (resp. backward selection)
+        all_features = set(range(X.shape[1]))
```

Comment:
_safe_indexing still supports boolean masks, right? That would make this easier, I think.

Reply:
Did you find this hard to read?

I like using sets because with boolean array we'd have to explicitly copy, and manually set values. No strong opinion.

I was also thinking that in the future we might want to expose a `selected_features_` attribute which for each iteration would contain the set of selected features at that stage.

Reply:
I find boolean arrays easier to read but I'm happy to leave it as it is.

Reply:
ok, using mask arrays now